File: src/page/info/filter/EnemyFilterBox.java
package page.info.filter;

import static utilpc.Interpret.ATKCONF;
import static utilpc.Interpret.EABI;
import static utilpc.Interpret.EABIIND;
import static utilpc.Interpret.EFILTER;
import static utilpc.Interpret.ERARE;
import static utilpc.Interpret.SPROC;
import static utilpc.Interpret.TRAIT;
import static utilpc.Interpret.isER;
import static utilpc.Interpret.isType;

import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.image.BufferedImage;
import java.util.ArrayList;
import java.util.List;
import java.util.Vector;

import javax.swing.AbstractButton;
import javax.swing.ImageIcon;
import javax.swing.JList;
import javax.swing.JScrollPane;
import javax.swing.ListSelectionModel;
import javax.swing.event.ListSelectionEvent;
import javax.swing.event.ListSelectionListener;

import common.system.MultiLangCont;
import common.util.pack.Pack;
import common.util.unit.Enemy;
import common.util.unit.EnemyStore;
import main.MainBCU;
import page.JTG;
import page.Page;
import utilpc.UtilPC;

public abstract class EnemyFilterBox extends Page {

	private static final long serialVersionUID = 1L;

	public static EnemyFilterBox getNew(Page p, Pack pack) {
		if (MainBCU.FILTER_TYPE == 0)
			return new EFBButton(p, pack);
		if (MainBCU.FILTER_TYPE == 1)
			return new EFBList(p, pack);
		return null;
	}

	protected Pack pac;
	
	protected String name = "";

	protected EnemyFilterBox(Page p, Pack pack) {
		super(p);
		pac = pack;
	}

	protected abstract int[] getSizer();

}

class EFBButton extends EnemyFilterBox {

	private static final long serialVersionUID = 1L;

	private final JTG[] orop = new JTG[3];
	private final JTG[] rare = new JTG[ERARE.length];
	private final JTG[] trait = new JTG[TRAIT.length];
	private final JTG[] abis = new JTG[EFILTER];
	private final JTG[] proc = new JTG[SPROC.length];
	private final JTG[] atkt = new JTG[ATKCONF.length];

	protected EFBButton(Page p, Pack pack) {
		super(p, pack);

		ini();
		confirm();
	}

	@Override
	protected int[] getSizer() {
		return new int[] { 2000, 400, 1, 400 };
	}

	@Override
	protected void resized(int x, int y) {
		JTG[][] btns = new JTG[][] { rare, trait, abis, proc, atkt };
		AttList.btnDealer(x, y, btns, orop, -1, 0, 1, -1, 2);
	}
	
	@Override
	public void callBack(Object o) {
		confirm();
	}


	private void confirm() {
		List<Enemy> ans = new ArrayList<>();
		for (Enemy e : EnemyStore.getAll(pac, true)) {
			int t = e.de.getType();
			int a = e.de.getAbi();
			boolean b0 = false;
			for (int i = 0; i < rare.length; i++)
				if (rare[i].isSelected())
					b0 |= isER(e, i);
			boolean b1 = !orop[0].isSelected();
			for (int i = 0; i < trait.length; i++)
				if (trait[i].isSelected())
					if (orop[0].isSelected())
						b1 |= ((t >> i) & 1) == 1;
					else
						b1 &= ((t >> i) & 1) == 1;
			boolean b2 = !orop[1].isSelected();
			for (int i = 0; i < abis.length; i++)
				if (abis[i].isSelected()) {
					boolean bind = ((a >> EABIIND[i]) & 1) == 1;
					if (orop[1].isSelected())
						b2 |= bind;
					else
						b2 &= bind;
				}
			for (int i = 0; i < proc.length; i++)
				if (proc[i].isSelected())
					if (orop[1].isSelected())
						b2 |= e.de.getAllProc().getArr(i).exists();
					else
						b2 &= e.de.getAllProc().getArr(i).exists();
			boolean b3 = !orop[2].isSelected();
			for (int i = 0; i < atkt.length; i++)
				if (atkt[i].isSelected())
					if (orop[2].isSelected())
						b3 |= isType(e.de, i);
					else
						b3 &= isType(e.de, i);
			boolean b4 = true;
			
			String ename;
			
			ename = MultiLangCont.ENAME.getCont(e);
			
			if(ename == null)
				ename = e.name;
			
			if(ename == null)
				ename = "";
			
			if(name != null) {
				b4 = ename.toLowerCase().contains(name.toLowerCase());
			}
			
			b0 = nonSele(rare) | b0;
			b1 = nonSele(trait) | b1;
			b2 = nonSele(abis) & nonSele(proc) | b2;
			b3 = nonSele(atkt) | b3;
			if (b0 & b1 & b2 & b3 & b4)
				ans.add(e);
		}
		getFront().callBack(ans);
	}

	private void ini() {
		for (int i = 0; i < orop.length; i++)
			set(orop[i] = new JTG(0, "orop"));
		for (int i = 0; i < rare.length; i++)
			set(rare[i] = new JTG(ERARE[i]));
		for (int i = 0; i < trait.length; i++) {
			set(trait[i] = new JTG(TRAIT[i]));
			BufferedImage v = UtilPC.getIcon(3, i);
			if (v == null)
				continue;
			trait[i].setIcon(new ImageIcon(v));
		}
		for (int i = 0; i < abis.length; i++) {
			set(abis[i] = new JTG(EABI[i]));
			BufferedImage v = UtilPC.getIcon(0, EABIIND[i]);
			if (v == null)
				continue;
			abis[i].setIcon(new ImageIcon(v));
		}
		for (int i = 0; i < proc.length; i++) {
			set(proc[i] = new JTG(SPROC[i]));
			BufferedImage v = UtilPC.getIcon(1, i);
			if (v == null)
				continue;
			proc[i].setIcon(new ImageIcon(v));
		}
		for (int i = 0; i < atkt.length; i++) {
			set(atkt[i] = new JTG(ATKCONF[i]));
			BufferedImage v = UtilPC.getIcon(2, i);
			if (v == null)
				continue;
			atkt[i].setIcon(new ImageIcon(v));
		}
	}

	private boolean nonSele(JTG[] jtbs) {
		int n = 0;
		for (int i = 0; i < jtbs.length; i++)
			if (jtbs[i].isSelected())
				n++;
		return n == 0;
	}

	private void set(AbstractButton b) {
		add(b);
		b.addActionListener(new ActionListener() {

			@Override
			public void actionPerformed(ActionEvent arg0) {
				confirm();
			}

		});
	}

}

class EFBList extends EnemyFilterBox {

	private static final long serialVersionUID = 1L;

	private final JTG[] orop = new JTG[3];
	private final JList<String> rare = new JList<>(ERARE);
	private final Vector<String> vt = new Vector<>();
	private final Vector<String> va = new Vector<>();
	private final AttList trait = new AttList(3, 0);
	private final AttList abis = new AttList(-1, EFILTER);
	private final AttList atkt = new AttList(2, 0);
	private final JScrollPane jr = new JScrollPane(rare);
	private final JScrollPane jt = new JScrollPane(trait);
	private final JScrollPane jab = new JScrollPane(abis);
	private final JScrollPane jat = new JScrollPane(atkt);

	protected EFBList(Page p, Pack pack) {
		super(p, pack);

		ini();
		confirm();
	}

	@Override
	protected int[] getSizer() {
		return new int[] { 450, 1150, 0, 500 };
	}

	@Override
	protected void resized(int x, int y) {
		set(orop[0], x, y, 0, 350, 200, 50);
		set(orop[1], x, y, 250, 0, 200, 50);
		set(orop[2], x, y, 0, 800, 200, 50);

		set(jr, x, y, 0, 50, 200, 250);
		set(jt, x, y, 0, 400, 200, 350);
		set(jab, x, y, 250, 50, 200, 1100);
		set(jat, x, y, 0, 850, 200, 300);
	}
	
	@Override
	public void callBack(Object o) {
		confirm();
	}

	private void confirm() {
		List<Enemy> ans = new ArrayList<>();
		for (Enemy e : EnemyStore.getAll(pac, true)) {
			int t = e.de.getType();
			int a = e.de.getAbi();
			boolean b0 = isER(e, rare.getSelectedIndex());
			boolean b1 = !orop[0].isSelected();
			for (int i : trait.getSelectedIndices())
				if (orop[0].isSelected())
					b1 |= ((t >> i) & 1) == 1;
				else
					b1 &= ((t >> i) & 1) == 1;
			boolean b2 = !orop[1].isSelected();
			int len = EFILTER;
			for (int i : abis.getSelectedIndices())
				if (i < len) {
					boolean bind = ((a >> EABIIND[i]) & 1) == 1;
					if (orop[1].isSelected())
						b2 |= bind;
					else
						b2 &= bind;
				} else if (orop[1].isSelected())
					b2 |= e.de.getAllProc().getArr(i - len).exists();
				else
					b2 &= e.de.getAllProc().getArr(i - len).exists();
			boolean b3 = !orop[2].isSelected();
			for (int i : atkt.getSelectedIndices())
				if (orop[2].isSelected())
					b3 |= isType(e.de, i);
				else
					b3 &= isType(e.de, i);
			
			boolean b4 = true;
			
			String ename;
			
			ename = MultiLangCont.ENAME.getCont(e);
			
			if(ename == null)
				ename = e.name;
			
			if(ename == null)
				ename = "";
			
			if(name != null) {
				b4 = ename.toLowerCase().contains(name.toLowerCase());
			}
			
			b0 = rare.getSelectedIndex() == -1 | b0;
			b1 = trait.getSelectedIndex() == -1 | b1;
			b2 = abis.getSelectedIndex() == -1 | b2;
			b3 = atkt.getSelectedIndex() == -1 | b3;
			if (b0 & b1 & b2 & b3 & b4)
				ans.add(e);
		}
		getFront().callBack(ans);
	}

	private void ini() {
		for (int i = 0; i < orop.length; i++)
			set(orop[i] = new JTG(get(0, "orop")));
		for (int i = 0; i < TRAIT.length; i++)
			vt.add(TRAIT[i]);
		for (int i = 0; i < EFILTER; i++)
			va.add(EABI[i]);
		for (int i = 0; i < SPROC.length; i++)
			va.add(SPROC[i]);
		trait.setListData(vt);
		abis.setListData(va);
		atkt.setListData(ATKCONF);
		rare.setSelectionMode(ListSelectionModel.SINGLE_SELECTION);
		int m = ListSelectionModel.MULTIPLE_INTERVAL_SELECTION;
		trait.setSelectionMode(m);
		abis.setSelectionMode(m);
		atkt.setSelectionMode(m);
		set(rare);
		set(trait);
		set(abis);
		set(atkt);
		add(jr);
		add(jt);
		add(jab);
		add(jat);
	}

	private void set(AbstractButton b) {
		add(b);
		b.addActionListener(new ActionListener() {

			@Override
			public void actionPerformed(ActionEvent arg0) {
				confirm();
			}

		});
	}

	private void set(JList<?> jl) {

		jl.addListSelectionListener(new ListSelectionListener() {

			@Override
			public void valueChanged(ListSelectionEvent arg0) {
				confirm();
			}

		});
	}

}

File: src/page/info/filter/EnemyFilterBox.kt
package page.info.filter

import common.io.assets.Admin
import common.io.assets.Admin.StaticPermitted
import common.io.assets.AssetLoader
import common.io.assets.AssetLoader.AssetHeader
import common.io.assets.AssetLoader.AssetHeader.AssetEntry
import common.io.json.JsonEncoder
import common.io.json.Test
import common.io.json.Test.JsonTest_0.JsonD
import common.io.json.Test.JsonTest_2
import common.pack.Source.AnimLoader
import common.pack.Source.ResourceLocation
import common.pack.Source.SourceAnimLoader
import common.pack.Source.SourceAnimSaver
import common.pack.Source.Workspace
import common.pack.Source.ZipSource
import common.pack.UserProfile
import common.util.lang.MultiLangCont
import common.util.stage.EStage
import common.util.stage.StageMap
import common.util.stage.StageMap.StageMapInfo
import common.util.unit.Enemy
import common.util.unit.UnitLevel
import io.BCPlayer
import main.MainBCU
import page.JL
import page.JTG
import page.Page
import page.anim.AnimBox
import page.info.filter.AttList
import page.support.ListJtfPolicy
import page.support.SortTable
import page.view.ViewBox
import page.view.ViewBox.Conf
import page.view.ViewBox.Controller
import page.view.ViewBox.VBExporter
import utilpc.Interpret
import utilpc.UtilPC
import java.awt.event.ActionEvent
import java.awt.event.ActionListener
import java.awt.image.BufferedImage
import java.util.*
import javax.swing.*
import javax.swing.event.ListSelectionEvent
import javax.swing.event.ListSelectionListener

abstract class EnemyFilterBox protected constructor(p: Page?, protected var pac: String?) : Page(p) {
    var name: String? = ""
    abstract fun getSizer(): IntArray

    companion object {
        private const val serialVersionUID = 1L
        fun getNew(p: Page?, pack: String?): EnemyFilterBox? {
            if (MainBCU.FILTER_TYPE == 0) return EFBButton(p, pack)
            return if (MainBCU.FILTER_TYPE == 1) EFBList(p, pack) else null
        }
    }
}

internal class EFBButton(p: Page?, pack: String?) : EnemyFilterBox(p, pack) {
    private val orop: Array<JTG?> = arrayOfNulls<JTG>(3)
    private val rare: Array<JTG?> = arrayOfNulls<JTG>(Interpret.ERARE.size)
    private val trait: Array<JTG?> = arrayOfNulls<JTG>(Interpret.TRAIT.size)
    private val abis: Array<JTG?> = arrayOfNulls<JTG>(Interpret.EFILTER)
    private val proc: Array<JTG?> = arrayOfNulls<JTG>(Interpret.SPROC.size)
    private val atkt: Array<JTG?> = arrayOfNulls<JTG>(Interpret.ATKCONF.size)
    override fun callBack(o: Any?) {
        confirm()
    }

    protected override fun getSizer(): IntArray {
        return intArrayOf(2000, 400, 1, 400)
    }

    override fun resized(x: Int, y: Int) {
        val btns: Array<Array<JTG?>> = arrayOf<Array<JTG?>>(rare, trait, abis, proc, atkt)
        AttList.Companion.btnDealer(x, y, btns, orop, -1, 0, 1, -1, 2)
    }

    private fun confirm() {
        val ans: MutableList<Enemy> = ArrayList<Enemy>()
        for (e in UserProfile.Companion.getAll<Enemy>(pac, Enemy::class.java)) {
            val t: Int = e.de.getType()
            val a: Int = e.de.getAbi()
            var b0 = false
            for (i in rare.indices) if (rare[i].isSelected()) b0 = b0 or Interpret.isER(e, i)
            var b1: Boolean = !orop[0].isSelected()
            for (i in trait.indices) if (trait[i].isSelected()) b1 = if (orop[0].isSelected()) b1 or (t shr i and 1 == 1) else b1 and (t shr i and 1 == 1)
            var b2: Boolean = !orop[1].isSelected()
            for (i in abis.indices) if (abis[i].isSelected()) {
                val bind = a shr Interpret.EABIIND.get(i) and 1 == 1
                b2 = if (orop[1].isSelected()) b2 or bind else b2 and bind
            }
            for (i in proc.indices) if (proc[i].isSelected()) b2 = if (orop[1].isSelected()) b2 or e.de.getAllProc().getArr(i).exists() else b2 and e.de.getAllProc().getArr(i).exists()
            var b3: Boolean = !orop[2].isSelected()
            for (i in atkt.indices) if (atkt[i].isSelected()) b3 = if (orop[2].isSelected()) b3 or Interpret.isType(e.de, i) else b3 and Interpret.isType(e.de, i)
            var b4 = true
            var ename: String
            ename = MultiLangCont.Companion.getStatic().ENAME.getCont(e)
            if (ename == null) ename = e.name
            if (ename == null) ename = ""
            if (name != null) {
                b4 = ename.toLowerCase().contains(name!!.toLowerCase())
            }
            b0 = nonSele(rare) or b0
            b1 = nonSele(trait) or b1
            b2 = nonSele(abis) and nonSele(proc) or b2
            b3 = nonSele(atkt) or b3
            if (b0 and b1 and b2 and b3 and b4) ans.add(e)
        }
        front.callBack(ans)
    }

    private fun ini() {
        for (i in orop.indices) set(JTG(0, "orop").also { orop[i] = it })
        for (i in rare.indices) set(JTG(Interpret.ERARE.get(i)).also { rare[i] = it })
        for (i in trait.indices) {
            set(JTG(Interpret.TRAIT.get(i)).also { trait[i] = it })
            val v: BufferedImage = UtilPC.getIcon(3, i) ?: continue
            trait[i].setIcon(ImageIcon(v))
        }
        for (i in abis.indices) {
            set(JTG(Interpret.EABI.get(i)).also { abis[i] = it })
            val v: BufferedImage = UtilPC.getIcon(0, Interpret.EABIIND.get(i)) ?: continue
            abis[i].setIcon(ImageIcon(v))
        }
        for (i in proc.indices) {
            set(JTG(Interpret.SPROC.get(i)).also { proc[i] = it })
            val v: BufferedImage = UtilPC.getIcon(1, i) ?: continue
            proc[i].setIcon(ImageIcon(v))
        }
        for (i in atkt.indices) {
            set(JTG(Interpret.ATKCONF.get(i)).also { atkt[i] = it })
            val v: BufferedImage = UtilPC.getIcon(2, i) ?: continue
            atkt[i].setIcon(ImageIcon(v))
        }
    }

    private fun nonSele(jtbs: Array<JTG?>): Boolean {
        var n = 0
        for (i in jtbs.indices) if (jtbs[i].isSelected()) n++
        return n == 0
    }

    private fun set(b: AbstractButton) {
        add(b)
        b.addActionListener(object : ActionListener {
            override fun actionPerformed(arg0: ActionEvent?) {
                confirm()
            }
        })
    }

    companion object {
        private const val serialVersionUID = 1L
    }

    init {
        ini()
        confirm()
    }
}

internal class EFBList(p: Page?, pack: String?) : EnemyFilterBox(p, pack) {
    private val orop: Array<JTG?> = arrayOfNulls<JTG>(3)
    private val rare: JList<String> = JList<String>(Interpret.ERARE)
    private val vt = Vector<String>()
    private val va = Vector<String>()
    private val trait = AttList(3, 0)
    private val abis = AttList(-1, Interpret.EFILTER)
    private val atkt = AttList(2, 0)
    private val jr: JScrollPane = JScrollPane(rare)
    private val jt: JScrollPane = JScrollPane(trait)
    private val jab: JScrollPane = JScrollPane(abis)
    private val jat: JScrollPane = JScrollPane(atkt)
    override fun callBack(o: Any?) {
        confirm()
    }

    protected override fun getSizer(): IntArray {
        return intArrayOf(450, 1150, 0, 500)
    }

    override fun resized(x: Int, y: Int) {
        Page.Companion.set(orop[0], x, y, 0, 350, 200, 50)
        Page.Companion.set(orop[1], x, y, 250, 0, 200, 50)
        Page.Companion.set(orop[2], x, y, 0, 800, 200, 50)
        Page.Companion.set(jr, x, y, 0, 50, 200, 250)
        Page.Companion.set(jt, x, y, 0, 400, 200, 350)
        Page.Companion.set(jab, x, y, 250, 50, 200, 1100)
        Page.Companion.set(jat, x, y, 0, 850, 200, 300)
    }

    private fun confirm() {
        val ans: MutableList<Enemy> = ArrayList<Enemy>()
        for (e in UserProfile.Companion.getAll<Enemy>(pac, Enemy::class.java)) {
            val t: Int = e.de.getType()
            val a: Int = e.de.getAbi()
            var b0: Boolean = Interpret.isER(e, rare.getSelectedIndex())
            var b1: Boolean = !orop[0].isSelected()
            for (i in trait.getSelectedIndices()) b1 = if (orop[0].isSelected()) b1 or (t shr i and 1 == 1) else b1 and (t shr i and 1 == 1)
            var b2: Boolean = !orop[1].isSelected()
            val len: Int = Interpret.EFILTER
            for (i in abis.getSelectedIndices()) b2 = if (i < len) {
                val bind = a shr Interpret.EABIIND.get(i) and 1 == 1
                if (orop[1].isSelected()) b2 or bind else b2 and bind
            } else if (orop[1].isSelected()) b2 or e.de.getAllProc().getArr(i - len).exists() else b2 and e.de.getAllProc().getArr(i - len).exists()
            var b3: Boolean = !orop[2].isSelected()
            for (i in atkt.getSelectedIndices()) b3 = if (orop[2].isSelected()) b3 or Interpret.isType(e.de, i) else b3 and Interpret.isType(e.de, i)
            var b4 = true
            var ename: String
            ename = MultiLangCont.Companion.getStatic().ENAME.getCont(e)
            if (ename == null) ename = e.name
            if (ename == null) ename = ""
            if (name != null) {
                b4 = ename.toLowerCase().contains(name!!.toLowerCase())
            }
            b0 = rare.getSelectedIndex() == -1 or b0
            b1 = trait.getSelectedIndex() == -1 or b1
            b2 = abis.getSelectedIndex() == -1 or b2
            b3 = atkt.getSelectedIndex() == -1 or b3
            if (b0 and b1 and b2 and b3 and b4) ans.add(e)
        }
        front.callBack(ans)
    }

    private fun ini() {
        for (i in orop.indices) set(JTG(Page.Companion.get(0, "orop")).also { orop[i] = it })
        for (i in Interpret.TRAIT.indices) vt.add(Interpret.TRAIT.get(i))
        for (i in 0 until Interpret.EFILTER) va.add(Interpret.EABI.get(i))
        for (i in Interpret.SPROC.indices) va.add(Interpret.SPROC.get(i))
        trait.setListData(vt)
        abis.setListData(va)
        atkt.setListData(Interpret.ATKCONF)
        rare.setSelectionMode(ListSelectionModel.SINGLE_SELECTION)
        val m: Int = ListSelectionModel.MULTIPLE_INTERVAL_SELECTION
        trait.setSelectionMode(m)
        abis.setSelectionMode(m)
        atkt.setSelectionMode(m)
        set(rare)
        set(trait)
        set(abis)
        set(atkt)
        add(jr)
        add(jt)
        add(jab)
        add(jat)
    }

    private fun set(b: AbstractButton) {
        add(b)
        b.addActionListener(object : ActionListener {
            override fun actionPerformed(arg0: ActionEvent?) {
                confirm()
            }
        })
    }

    private fun set(jl: JList<*>) {
        jl.addListSelectionListener(object : ListSelectionListener {
            override fun valueChanged(arg0: ListSelectionEvent?) {
                confirm()
            }
        })
    }

    companion object {
        private const val serialVersionUID = 1L
    }

    init {
        ini()
        confirm()
    }
}
